test/hotspot/jtreg/runtime/vthread/JNIMonitor/JNIMonitor.java

```diff
@@ -21,42 +21,278 @@
+import java.lang.reflect.InvocationTargetException;
+import java.util.List;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/*
+ * Tests that JNI monitors work correctly with virtual threads,
+ * There are multiple test scenarios that we check using unified logging output
+ * (both positive and negative tests). Each test case is handled by its own @-test
+ * definition so that we can run each sub-test independently.
+ *
+ * The original bug was only discovered because the ForkJoinPool worker thread terminated
+ * and trigerred an assertion failure. So we use a custom scheduler to give us control.
+ */
 
 /**
- * @test JNIMonitor
- * @summary Tests that JNI monitors work correctly with virtual threads
+ * @test id=normal
+ * @bug 8327743
+ * @summary Normal lock then unlock
  * @library /test/lib
- * @compile JNIMonitor.java
- * @run main/native/othervm JNIMonitor
+ * @modules java.base/java.lang:+open
+ * @requires vm.continuations
+ * @run driver JNIMonitor Normal
+ */
+
+/**
+ * @test id=multiNormal
+ * @bug 8327743
+ * @summary Normal lock then unlock by multiple threads
+ * @library /test/lib
+ * @modules java.base/java.lang:+open
+ * @requires vm.continuations
+ * @run driver JNIMonitor MultiNormal
+ */
+
+/**
+ * @test id=missingUnlock
+ * @bug 8327743
+ * @summary Don't do the unlock and exit normally
+ * @library /test/lib
+ * @modules java.base/java.lang:+open
+ * @requires vm.continuations
+ * @run driver JNIMonitor MissingUnlock
+ */
+
+/**
+ * @test id=multiMissingUnlock
+ * @bug 8327743
+ * @summary Don't do the unlock and exit normally, by multiple threads
+ * @library /test/lib
+ * @modules java.base/java.lang:+open
+ * @requires vm.continuations
+ * @run driver JNIMonitor MultiMissingUnlock
+ */
+
+/**
+ * @test id=missingUnlockWithThrow
+ * @bug 8327743
+ * @summary Don't do the unlock and exit by throwing
+ * @library /test/lib
+ * @modules java.base/java.lang:+open
+ * @requires vm.continuations
+ * @run driver JNIMonitor MissingUnlockWithThrow
+ */
+
+/**
+ * @test id=multiMissingUnlockWithThrow
+ * @bug 8327743
+ * @summary Don't do the unlock and exit by throwing, by multiple threads
+ * @library /test/lib
+ * @modules java.base/java.lang:+open
+ * @requires vm.continuations
+ * @run driver JNIMonitor MultiMissingUnlockWithThrow
  */
 
 public class JNIMonitor {
 
+    public static void main(String[] args) throws Exception {
+        String test = args[0];
+        String[] cmdArgs = new String[] {
+            "-Djava.library.path=" + Utils.TEST_NATIVE_PATH,
+            // The following is a hack to trick the pool worker threads into terminating
+            // after one second (default keep-alive / parallelism).
+            "-Djdk.virtualThreadScheduler.parallelism=30",
```

Comment:
Leftover from old testing?

Reply:
Yep - well spotted. Fixed. Also fixed incorrect comment about the add-opens.